grakn-graql/src/main/java/ai/grakn/graql/internal/gremlin/fragment/OutShortcutFragment.java

```diff
@@ -46,36 +47,45 @@
 class OutShortcutFragment extends AbstractFragment {
 
     private final Var edge;
-    private final Optional<Set<TypeLabel>> roleTypes;
-    private final Optional<Set<TypeLabel>> relationTypes;
+
+    private final Optional<Var> roleType;
+    private final Optional<Set<TypeLabel>> roleTypeLabels;
+    private final Optional<Set<TypeLabel>> relationTypeLabels;
 
     OutShortcutFragment(
-            Var relation, Var edge, Var rolePlayer, Optional<Set<TypeLabel>> roleTypes,
-            Optional<Set<TypeLabel>> relationTypes) {
-            super(relation, rolePlayer, edge);
+            Var relation, Var edge, Var rolePlayer, Optional<Var> roleType, Optional<Set<TypeLabel>> roleTypeLabels,
+            Optional<Set<TypeLabel>> relationTypeLabels) {
+            super(relation, rolePlayer, roleType.map(rt -> new Var[] {edge, rt}).orElse(new Var[] {edge}));
```

Comment:
I would have gone with creating a new constructor. This is horribly difficult to read. Just a suggestion though.